tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/PopulateBasicPropertiesHeadersIntegration.cs

```diff
@@ -45,8 +53,35 @@ public class PopulateBasicPropertiesHeadersIntegration
             return new CallTargetReturn<TReturn?>(returnValue);
         }
 
-        returnValue ??= CachedBasicPropertiesHelper<TReturn>.CreateHeaders();
-        var duckType = returnValue.DuckCast<IBasicProperties>()!;
+        TReturn? basicProperties = default;
+
+        // PopulateBasicPropertiesHeaders returns null if the supplied IReadOnlyBasicProperties
+        // does not have to be modified or if it's a writable instance.
+        // If that is the case then we have to fetch IReadOnlyBasicProperties from the argument
+        // list instead of creating a new instance that overwrites the supplied properties.
+        if (returnValue is null)
+        {
+            var rabbitMqClientAssembly = typeof(TReturn).Assembly;
+
+            if (state.State?.GetType() == rabbitMqClientAssembly.GetType("RabbitMQ.Client.BasicProperties", throwOnError: false))
+            {
+                // Use the existing basic properties if possible...
+                basicProperties = (TReturn)state.State!;
+            }
+            else if (state.State?.GetType() == rabbitMqClientAssembly.GetType("RabbitMQ.Client.IReadOnlyBasicProperties", throwOnError: false))
+            {
+                // if not create new instance using the copy constructor on BasicProperties.
+                basicProperties = CachedBasicPropertiesHelper<TReturn>.CreateHeaders(state.State!);
+            }
+            else
+            {
+                // This case cannot happen as argument is not nullable.
+                Log.Warning("Invalid state: PopulateBasicPropertiesHeaders() is returning null and CallTargetState.State has type {Type}", state.State?.GetType().FullName ?? "null");
+                return new CallTargetReturn<TReturn?>(returnValue);
+            }
```

Comment:
I _think_ if we invert these we can avoid the ugly `GetType()`? 🤔 
```suggestion
            else if (state.State is null)
            {
                // This case cannot happen as argument is not nullable.
                Log.Warning("Invalid state: PopulateBasicPropertiesHeaders() is returning null and CallTargetState.State has type {Type}", state.State?.GetType().FullName ?? "null");
                return new CallTargetReturn<TReturn?>(returnValue);
            }
            else
            {
                // if not create new instance using the copy constructor on BasicProperties.
                basicProperties = CachedBasicPropertiesHelper<TReturn>.CreateHeaders(state.State!);
            }
```